dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/document/visibility/DocumentVisibility.java

```diff
@@ -0,0 +1,584 @@
+          if (diff != 0) {
+			return diff;
+		}
+          for (int i = 0; i < a.children.size(); i++) {
+            diff = compare(a.children.get(i), b.children.get(i));
+            if (diff != 0) {
+				return diff;
+			}
+          }
+      }
+      return 0;
+    }
+  }
+
+  /*
+   * Convience method that delegates to normalize with a new NodeComparator constructed using the supplied expression.
+   */
+  public static Node normalize(final Node root, final byte[] expression) {
+    return normalize(root, expression, new NodeComparator(expression));
+  }
+
+  // @formatter:off
+  /*
+   * Walks an expression's AST in order to:
+   *  1) roll up expressions with the same operant (`a&(b&c) becomes a&b&c`)
+   *  2) sorts labels lexicographically (permutations of `a&b&c` are re-ordered to appear as `a&b&c`)
+   *  3) dedupes labels (`a&b&a` becomes `a&b`)
+   */
+  // @formatter:on
+  public static Node normalize(final Node root, final byte[] expression, final NodeComparator comparator) {
+    if (root.type != NodeType.TERM) {
+      final TreeSet<Node> rolledUp = new TreeSet<>(comparator);
+      final java.util.Iterator<Node> itr = root.children.iterator();
+      while (itr.hasNext()) {
+        final Node c = normalize(itr.next(), expression, comparator);
+        if (c.type == root.type) {
+          rolledUp.addAll(c.children);
+          itr.remove();
+        }
+      }
+      rolledUp.addAll(root.children);
+      root.children.clear();
+      root.children.addAll(rolledUp);
+
+      // need to promote a child if it's an only child
+      if (root.children.size() == 1) {
+        return root.children.get(0);
+      }
+    }
+
+    return root;
+  }
+
+  /*
+   * Walks an expression's AST and appends a string representation to a supplied StringBuilder. This method adds parens where necessary.
+   */
+  public static void stringify(final Node root, final byte[] expression, final StringBuilder out) {
+    if (root.type == NodeType.TERM) {
+      out.append(new String(expression, root.start, root.end - root.start, UTF_8));
+    } else {
+      String sep = "";
+      for (final Node c : root.children) {
+        out.append(sep);
+        final boolean parens = (c.type != NodeType.TERM && root.type != c.type);
+        if (parens) {
+			out.append("(");
+		}
+        stringify(c, expression, out);
+        if (parens) {
+			out.append(")");
+		}
+        sep = root.type == NodeType.AND ? "&" : "|";
+      }
+    }
+  }
+
+  /**
+   * Generates a byte[] that represents a normalized, but logically equivalent, form of this evaluator's expression.
+   *
+   * @return normalized expression in byte[] form
+   */
+  public byte[] flatten() {
+    final Node normRoot = normalize(node, expression);
+    final StringBuilder builder = new StringBuilder(expression.length);
+    stringify(normRoot, expression, builder);
+    return builder.toString().getBytes(UTF_8);
+  }
+
+  private static class DocumentVisibilityParser {
```

Comment:
this needs to be heavily documented.  Its not very clear how this is getting parsed, from what to what